TEP-0009: Trigger CRD (#148)
tektoncd/community: teps/0008-trigger-crd.md

### Non-Goals

How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
This will be handled by operator or admin of the cluster.

## Requirements

1. EventListener can create Tekton resources in a single namespace or across multiple namespaces if configured to do so.
2. Binding of these cross namespace Trigger to EventListener. When an EL recieves an event, it can process configured subset of triggers using path/selector.
3. EL can still process triggers defined in the current form aka backwards compatibility.

## Proposal

EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.

### Trigger CRD
A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
a single TriggerTemplate, and optionally interceptors and a serviceAccount.
```
apiVersion: v1alpha1
kind: Trigger
metadata:
  name: my-repo-trigger
  labels:
    eventlistener: operator
spec:
  serviceAccountName: "blah"
  interceptors:
    cel:
      filter: "$(header.eventType == "push")"

>   bindings:

Reviewer: The binding syntax here is the one I proposed in https://github.com/tektoncd/triggers/issues/617

I think we should use the one that we actually support for now 😄

Author: Good call. Fixed that.

    - name: key
      value: "$(body.commit_id)"
  template: "ref-to-my-template"
 status:
   address:
      url: "el-my-svc.cluster.local" # could also be an IP address

>    conditions:

Reviewer: conditions need a Type. I think the type here should be `Ready` which along with `Succeeded` are two of the "standard" condition types

Author: Done

are directed and process those requests with configuration from bound Triggers. EventListener has
triggerSelector to select triggers based on labels and namespaceSelector to select namespaces where
it searches for triggers.
```
apiVersion: v1alpha1
kind: EventListener
metadata:
  name: my-el
spec:
  serviceAccountName: "blah"
  namespaceSelector:
     matchName: [ns-1, ns-2]
     matchLabels:
       - tekton-triggers: tekton-operator
  triggerSelector:
    matchLabels:
      - eventlistener: tekton-operator
```

### Path based EventListener
Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
To refer to trigger resource ```foo``` in namespace ```bar```, we would have a url ```/bar?name=foo```.
Instead of EventListener iterating through every Triggers that can be serve, EventListener
will directly process the Trigger refer by URL. Similarly we can have url ```/bar?label=app-foo```
where EventListener will only process all the Triggers matching labels  ```app-foo```.

### User Stories (optional)

We have two type of users:
1. End users who requires his event to be processed.
2. Operator/admin who managed EventListener.

> #### End User handling Webhook Use Cases using TriggerCR

Reviewer 2: Still not using the classic user story wording of "As a <user persona>, I want ..." 

But I won't block on that if everybody else is ok with it.  The separation of concerns that is at the heart of the user stories and different roles is written down / captured in some form.

spec:
  serviceAccountName: "blah"
  namespaceSelector:
     matchName: [ns-1, ns-2]
     matchLabels:
       - triggers: tekton-operator
  triggerSelector:
    matchLabels:
      - eventlistener: tekton-operator
```

Trigger CR will look like the below example:
```
apiVersion: v1alpha1
kind: Trigger
metadata:
  name: my-repo-trigger
  labels:
    eventlistener: tekton-operator
spec:
  serviceAccountName: "blah"
  interceptors:
    cel:
      filter: "$(header.eventType == "push")"
  bindings:
    - name: key
      value: "$(body.commit_id)"
  template: "ref-to-my-template"
 status:
   address:
      url: "el-my-svc.cluster.local" # could also be an IP address?
   conditions:
      status: True
      message: "Bound to EventListener my-el"
```

Next, we will introduce Path base EventListener. Here instead of iterating through all
the triggers, EventListener will directly process the Trigger refer by the webhook URL. If we webhook
URL ```/bar?label=app-foo```,  EventListener will only process all the Triggers matching labels
```app-foo```.

Triggers will have labels which is used by `triggerSelector` of EventListener. This will be implemented
after path base EventListener.

> ## Alternatives

Reviewer: There are quite a few other alternatives that we discussed in regard to Selectors:

https://docs.google.com/document/d/1zWpmEhtSNe8KAPKvTJE7Pjg5Uzk8mx9MUjSN24D1jUg/edit#heading=h.6b7nnc6nrh2t

Probably worth calling that out here.

Author: Added that.

---
title: trigger-crd
authors:
  - "@kbaig"
creation-date: 2020-07-14
last-updated: 2020-07-14
status: proposed
---
<!--
**Note:** When your TEP is complete, all of these comment blocks should be removed.

To get started with this template:

- [ ] **Fill out this file as best you can.**
  At minimum, you should fill in the "Summary", and "Motivation" sections.
  These should be easy if you've preflighted the idea of the TEP with the
  appropriate Working Group.
- [ ] **Create a PR for this TEP.**
  Assign it to people in the SIG that are sponsoring this process.
- [ ] **Merge early and iterate.**
  Avoid getting hung up on specific details and instead aim to get the goals of
  the TEP clarified and merged quickly.  The best way to do this is to just
  start with the high-level sections and fill out details incrementally in

>   subsequent PRs.

Reviewer: can we remove these boilerplate comments?

Author: Done.

Just because a TEP is merged does not mean it is complete or approved.  Any TEP
marked as a `proposed` is a working document and subject to change.  You can
denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing TEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused.  If you disagree with what is already in a document, open a new PR
with suggested changes.

If there are new details that belong in the TEP, edit the TEP.  Once a
feature has become "implemented", major changes should get new TEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/teps/NNNN-TEP-template/README.md).

-->

> # TEP-0008: Introducing TriggerCRD

Reviewer: 0008 seems to be already taken by Knative EventListener. Use 0009?

Author: Which ever is merged first would get that number.

Reviewer 3: NIT: the number does not have to be consistent with the order of PR merged, but I don't mind either way

Author: OK. Added it as 9.

<!--
This is the title of your TEP.  Keep it short, simple, and descriptive.  A good
title can help communicate what the TEP is and should be considered as part of
any review.
-->

<!--
A table of contents is helpful for quickly jumping to sections of a TEP and for
highlighting any additional information provided beyond the standard TEP
template.

Ensure the TOC is wrapped with
  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
tags, and then generate with `hack/update-toc.sh`.
-->

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Requirements](#requirements)
- [Proposal](#proposal)
  - [Trigger CRD](#trigger-crd)
    - [Trigger Status](#trigger-status)
  - [EventListener CRD](#eventlistener-crd)
  - [Path base EventListener CRD](#path-base-eventlistener)
  - [User Stories (optional)](#user-stories-optional)

>     - [Story 1](#story-1)

Reviewer: dead links. remove

2. Ability to have single EventListener to cater to whole cluster.
3. Separation of Concerns :  The more ELs there are, the more additional configuration are needed - in some cases this additional bit of configuration is something that a cluster admin/operator would perform not an user/app developer.
For declarative config as code scenarios, it would be easier for a user to have create the Trigger configuration part as part of config with code. The cluster admin/operator can create the EventListeners, expose it to a public address, etc. And the same listener can be shared among many Triggers. This does reduce the number of listener Pods but also allows for a better separation of concerns among the user and the operator.

### Non-Goals

<!--
What is out of scope for this TEP?  Listing non-goals helps to focus discussion
and make progress.
-->
How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
This will be handled by operator or admin of the cluster.

## Requirements

<!--
List the requirements for this TEP.
-->
1. Can create Tekton resources in a single namespace or across multiple namespaces if configured to do so.

> 2. When an EL recieves an event, it can process only some configured subset of triggers <- path/selector based

Reviewer 2: This feels like it needs some further details / clarification.  Are we explicitly reducing the amount of trigger path/selector and EventListener can process today when there is a one to one mapping between EL and triggers?  If so, why is this needed?

Reviewer 2: If this is suppose to be alluding to the `Path based EventListener` section below, it is not obvious to the uninitiated

Author: I was talking about binding Trigger to EventListener here. Updating it.

3. EL can still process triggers defined in the current form aka backwards compatibility.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation.  The "Design Details" section below is for the real
nitty-gritty.
-->
EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.

### Trigger CRD
A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
a single TriggerTemplate, and optionally interceptors and a serviceAccount. 
```
apiVersion: v1alpha1
kind: Trigger
metadata:
  name: my-repo-trigger
  labels:
    eventlistener: operator
spec:
  serviceAccountName: "blah"
  interceptors:
    cel:
      filter: "$(header.eventType == "push")"
  bindings:
    - name: key
      value: "$(body.commit_id)"
  template: "ref-to-my-template"
 status:
   address:

>       url: "el-my-svc.cluster.local" # could also be an IP address?

Reviewer 2: Indicated by some of the questions/comments below, some detail in this section on what *exactly* this URL gets the Trigger author, etc. is needed.

Is it simply and identifier?  Is it the URL the trigger author configures in their SCM (github/gitlab, etc) webhook configuration?  If the trigger author does a curl against that URL, what should they expect?

kind: EventListener
metadata:
  name: my-el
spec:
  serviceAccountName: "blah"
  namespaceSelector:
     matchName: [ns-1, ns-2]
     matchLabels:
       - tekton-triggers: tekton-operator
  triggerSelector:
    matchLabels:
      - eventlistener: tekton-operator
```

### Path based EventListener
Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
To refer to trigger resource ```foo``` in namespace ```bar```, we would have a url ```/bar?name=foo```.
Instead of EventListener iterating through every Triggers that can be serve, EventListener
will directly process the Trigger refer by URL. Similarly we can have url ```/bar?label=app-foo``` 
where EventListener will only process all the Triggers matching labels  ```app-foo```.

### User Stories (optional)

<!--
Detail the things that people will be able to do if this TEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system.  The goal here is to make this feel real for users without getting
bogged down.
-->
We have two type of users:
1. End users who requires his event to be processed.
2. Operator/admin who managed EventListener.

> #### Trigger Ref inside the EventListener

Reviewer: we can call this section end users and describe how it works for end users? In addition to what you already have, maybe we can add how a end user goes about using it once a trigger is created i.e. getting listener address from status, setting up webhook of the form  http(s)://url/namespace/trigger . Also, how they can find the logs - find listener pod and then get logs for those?

Reviewer 2: Yeah I would have expected something like "As an end user, I am only concerned with defining the Trigger* objects in my namespace in order to have my SCM system trigger my Tekton pipielines, and do not want to have to manage the event listener sink myself."

And then "As a cluster administrator, I want to be able to manage the ingress points into my cluster's various use of tekton triggers, minimizing the amount of EventListerners/Sinks needed, and thus be able to absolve my end users from managing these http endpoints."

Author: I updated this section.

Instead of defining Trigger inside EventListener, Trigger resource will be created by end user.
Trigger will contain serviceaccount, triggerbinding, triggertemplate and interceptor.ServiceAccount
will be optional. If not defined, ServiceAccount of EventListener will be used.

> #### Specifying Namespaces where EventListener can serve

Reviewer: Similar to above, this section looks like that the user story would be for an operator user.  We could describe that an operator can create one or more listeners and then use a combination of namespace and trigger selectors to "select" which triggers/namespaces to serve. They can expose the Listener via Ingress/LoadBalancer and can retain write access to the Listener objects. They can choose to restrict write access to Listener objects while allowing users with read access to use the Listener/get logs etc.

Reviewer 2: +1 on the user story feel .... "As a cluster administrator, I want ot control which namespaces the HTTP endpoints needed for tekton triggers reside in".

Author: I updated this section.

Admin/Operator would give which namespaces EventListener can serve. In this way, EventListener can be used
in Multitenant Scenarios.

> #### Path Base EventListener Implementation

Reviewer: Is this supposed to be part of the user stories section?

Author: Removed it.

Path base EventListener will be implemented.

### Notes/Constraints/Caveats (optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above.
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations
<!--
What are the risks of this proposal and how do we mitigate. Think broadly.
For example, consider both security and how this will impact the larger
kubernetes ecosystem.

How will security be reviewed and by whom?

How will UX be reviewed and by whom?

Consider including folks that also work outside the WGs or subproject.
-->

> 1. Security issue - Wide permission to EventListener. In this scenario EventListener

Reviewer: can you describe a bit more detail what the wide permissions are? How can an admin manage it e.g. they may want to restrict write access to the Listener

Reviewer 2: Yes some elaboration would be good here.  i.e. that a single event listener would either have to have permissions to each of the namespaces, or for convenience sake, get clusterrole permissions, but that can give the EL access to "system level namesapces", like where the apiserver live, if done naively.

Or that a user could purposefully configure a github webhook to get to a "restricted" namesapce instead of a namespace they had access to if they got a hold of the verification secret, etc.

Stuff along those lines.

Author: I rephrased and added how it will be mitigated. I was talking about EventListener having wide permissions,

will function in the same way as controller. It will be the responsibility of Operator
or admin to manage that.

> 2. Log: How will user access the EventListener log to debug their events. This can be addressed via

Reviewer: The trigger status should contain the listeners right? And users can the get the listener pod and read its logs?

Reviewer 2: The trigger status has the URL, but that does not necessarily mean they have permission to dump pod logs in the EL namespace.

Author: Yes, that's correct. I believe EventListener creating events when processing Triggers is the feasible approach for this.

> emitting kubernetes events by EventListener and tkn-cli.

Reviewer 2: I know in WG we discussed some form of trigger specific log snippet from the EL log into the trigger status ... did you decide to abandon this @khrm ?

Author: Yes, we decided to abandon it. It might require another CRD itself.

3. Resource Hog issue: A particular namespace or project could hog most of the Resources. Operator
can handle this issue. This is similar to triggers controller or webhook. Operator can parition heavy user to its own EL. Also, ELs are stateless and should be horizontally scalable.

> ### User Experience (optional)

Reviewer: Let's either fill out this section or remove it

Author: I will add a ux  flow tomorrow.

<!--
Consideration about the user experience. Depending on the area of change,
users may be task and pipeline editors, they may trigger task and pipeline
runs or they may be responsible for monitoring the execution of runs,
via CLI, dashboard or a monitoring system.

Consider including folks that also work on CLI and dashboard.
-->

### Performance (optional)

> Path base EventListener will improve the performance of EventListener by directly targetting

Reviewer: ```suggestion
Path based EventListeners will improve the performance of EventListener by directly targeting
```

> the trigger definition we want to target instead of needlessly processing trigger that we don't.

Reviewer: ```suggestion
the trigger definition we want to target instead of processing each trigger within a EventListener like we do today.

> At the same time, one poorly behaved customer/namespace could dominate an eventlistener and effect

Reviewer: ```suggestion
At the same time, one poorly behaved customer/namespace could dominate an EventListener and affect
```

other namespaces, which wouldn't be the case if each namespace had their own eventlistener.
Operator/Admin can handle this issue by horizontally scaling EventListener or partitioning EventListener.
<!--
Consideration about performance.
What impact does this change have on the start-up time and execution time
of task and pipeline runs? What impact does it have on the resource footprint
of Tekton controllers as well as task and pipeline runs?

Consider which use cases are impacted by this change and what are their
performance requirements.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable.  This may include API specs (though not always
required) or even code snippets.  If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->
In first phases, this involve introduction of TriggerCRD and  modification of
EventListenerCRD to introduce selectors which searches for these Trigger resource.

> 

Reviewer: My understanding is:
1. Introduce Trigger CRD and let EL's use them via Reference
2. Add selectors (maybe only namespace names first)
3. Path based urls for triggers to Listeners
3. Add other selectors? labels

Author: Yes, done that now.
